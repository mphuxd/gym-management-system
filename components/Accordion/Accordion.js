import React from "react";
import * as AccordionPrimitive from "@radix-ui/react-accordion";

const Accordion = React.forwardRef(({ children, ...props }, forwardedRef) => {
  return (
    <AccordionPrimitive.Root ref={forwardedRef} {...props}>
      {children}
    </AccordionPrimitive.Root>
  );
});

Accordion.displayName = "Accordion";

export default Accordion;
