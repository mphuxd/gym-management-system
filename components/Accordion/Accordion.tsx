import React from 'react';
import * as AccordionPrimitive from '@radix-ui/react-accordion';

export const ACCORDION_TEST_ID: string = 'accordion';

type RootElement = React.ElementRef<typeof AccordionPrimitive.Root>;
type Props = React.ComponentPropsWithoutRef<
  typeof AccordionPrimitive.Root
> & {};

const Accordion = React.forwardRef<RootElement, Props>(
  ({ children, ...props }, forwardedRef) => (
    <AccordionPrimitive.Root
      data-testid={ACCORDION_TEST_ID}
      ref={forwardedRef}
      {...props}
    >
      {children}
    </AccordionPrimitive.Root>
  )
);

export default Accordion;
