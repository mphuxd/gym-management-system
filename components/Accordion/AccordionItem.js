import React from 'react';
import * as AccordionPrimitive from '@radix-ui/react-accordion';
import { ChevronDown } from '@carbon/icons-react';
import cx from 'classnames';
import styles from './Accordion.module.scss';

const AccordionItem = React.forwardRef(
  ({ children, header, ...props }, forwardedRef) => {
    const classNames = cx(
      'flex flex-row justify-between items-center w-full',
      styles.AccordionTrigger
    );
    return (
      <AccordionPrimitive.Item ref={forwardedRef} {...props}>
        <AccordionPrimitive.Header>
          <AccordionPrimitive.Trigger className={classNames}>
            <h3>{header}</h3>
            <ChevronDown className={styles.AccordionChevron} />
          </AccordionPrimitive.Trigger>
        </AccordionPrimitive.Header>
        <AccordionPrimitive.Content>{children}</AccordionPrimitive.Content>
      </AccordionPrimitive.Item>
    );
  }
);

AccordionItem.displayName = 'AccordionItem';

export default AccordionItem;
