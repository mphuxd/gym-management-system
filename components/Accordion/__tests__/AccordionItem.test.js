import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Accordion, { ACCORDION_TEST_ID } from '../Accordion';
import AccordionItem, {
  ACCORDION_CONTENT_TEST_ID,
  ACCORDION_HEADER_TEST_ID,
  ACCORDION_ITEM_TEST_ID,
  ACCORDION_TRIGGER_TEST_ID,
} from '../AccordionItem';

describe('AccordionItem: ', () => {
  it('should render', async () => {
    render(
      <Accordion>
        <AccordionItem header="item-title">content</AccordionItem>
      </Accordion>
    );
    const item = await screen.findByTestId(ACCORDION_ITEM_TEST_ID);
    expect(item).toMatchSnapshot();
    const header = await screen.findByTestId(ACCORDION_HEADER_TEST_ID);
    expect(header).toMatchSnapshot();
    const content = await screen.findByTestId(ACCORDION_CONTENT_TEST_ID);
    expect(content).toMatchSnapshot();
    const trigger = await screen.findByTestId(ACCORDION_TRIGGER_TEST_ID);
    expect(trigger).toMatchSnapshot();
  });

  it('should pass string to header', async () => {
    const headerString = 'item-header';
    render(
      <Accordion>
        <AccordionItem header={headerString}>content</AccordionItem>
      </Accordion>
    );
    const header = await screen.findByTestId(ACCORDION_HEADER_TEST_ID);
    expect(header).toHaveTextContent(headerString);
  });

  it('should be collapsed and hide content on click', async () => {
    const className = 'class-test';
    render(
      <Accordion
        type="single"
        defaultValue="test"
        collapsible
        className={className}
      >
        <AccordionItem value="test" header="Section 1 title">
          content
        </AccordionItem>
      </Accordion>
    );
    const buttons = await screen.findAllByTestId(ACCORDION_TRIGGER_TEST_ID);
    expect(screen.queryByText('content')).toBeInTheDocument();
    expect(await buttons[0]).toHaveAttribute('data-state', 'open');
    await userEvent.click(buttons[0]);
    expect(screen.queryByText('content')).not.toBeInTheDocument();
    expect(await buttons[0]).toHaveAttribute('data-state', 'closed');
    const accordion = await screen.findByTestId(ACCORDION_TEST_ID);
    expect(accordion).toMatchSnapshot();
  });

  it('should be expanded and show content on click', async () => {
    const className = 'class-test';
    render(
      <Accordion
        type="single"
        defaultValue="test"
        collapsible
        className={className}
      >
        <AccordionItem value="test" header="Section 1 title">
          content
        </AccordionItem>
      </Accordion>
    );
    const buttons = await screen.findAllByTestId(ACCORDION_TRIGGER_TEST_ID);
    expect(screen.queryByText('content')).toBeInTheDocument();
    expect(await buttons[0]).toHaveAttribute('data-state', 'open');
    await userEvent.click(buttons[0]);
    expect(screen.queryByText('content')).not.toBeInTheDocument();
    expect(await buttons[0]).toHaveAttribute('data-state', 'closed');
    await userEvent.click(buttons[0]);
    expect(screen.queryByText('content')).toBeInTheDocument();
    expect(await buttons[0]).toHaveAttribute('data-state', 'open');
    const accordion = await screen.findByTestId(ACCORDION_TEST_ID);
    expect(accordion).toMatchSnapshot();
  });
});
