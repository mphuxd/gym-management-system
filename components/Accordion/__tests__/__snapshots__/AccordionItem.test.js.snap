// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItem:  should be collapsed and hide content on click 1`] = `
<div
  class="class-test"
  data-testid="accordion"
>
  <div
    data-state="closed"
    data-testid="accordion-item"
  >
    <h3
      data-state="closed"
      data-testid="accordion-header"
    >
      <button
        aria-controls="radix-:r5:"
        aria-expanded="false"
        class="group flex w-full flex-row items-center justify-between bg-transparent"
        data-radix-collection-item=""
        data-state="closed"
        data-testid="accordion-trigger"
        id="radix-:r4:"
        type="button"
      >
        <h3>
          Section 1 title
        </h3>
        <svg
          aria-hidden="true"
          class="transition duration-300 group-data-[state=open]:rotate-180 group-data-[state=open]:transform"
          fill="currentColor"
          focusable="false"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
          />
        </svg>
      </button>
    </h3>
    <div
      aria-labelledby="radix-:r4:"
      data-state="closed"
      data-testid="accordion-content-text"
      hidden=""
      id="radix-:r5:"
      role="region"
      style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
    />
  </div>
</div>
`;

exports[`AccordionItem:  should be expanded and show content on click 1`] = `
<div
  class="class-test"
  data-testid="accordion"
>
  <div
    data-state="open"
    data-testid="accordion-item"
  >
    <h3
      data-state="open"
      data-testid="accordion-header"
    >
      <button
        aria-controls="radix-:r7:"
        aria-expanded="true"
        class="group flex w-full flex-row items-center justify-between bg-transparent"
        data-radix-collection-item=""
        data-state="open"
        data-testid="accordion-trigger"
        id="radix-:r6:"
        type="button"
      >
        <h3>
          Section 1 title
        </h3>
        <svg
          aria-hidden="true"
          class="transition duration-300 group-data-[state=open]:rotate-180 group-data-[state=open]:transform"
          fill="currentColor"
          focusable="false"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
          />
        </svg>
      </button>
    </h3>
    <div
      aria-labelledby="radix-:r6:"
      data-state="open"
      data-testid="accordion-content-text"
      id="radix-:r7:"
      role="region"
      style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
    >
      content
    </div>
  </div>
</div>
`;

exports[`AccordionItem:  should render 1`] = `
<div
  data-state="closed"
  data-testid="accordion-item"
>
  <h3
    data-state="closed"
    data-testid="accordion-header"
  >
    <button
      aria-controls="radix-:r1:"
      aria-expanded="false"
      class="group flex w-full flex-row items-center justify-between bg-transparent"
      data-radix-collection-item=""
      data-state="closed"
      data-testid="accordion-trigger"
      id="radix-:r0:"
      type="button"
    >
      <h3>
        item-title
      </h3>
      <svg
        aria-hidden="true"
        class="transition duration-300 group-data-[state=open]:rotate-180 group-data-[state=open]:transform"
        fill="currentColor"
        focusable="false"
        height="16"
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
        />
      </svg>
    </button>
  </h3>
  <div
    aria-labelledby="radix-:r0:"
    data-state="closed"
    data-testid="accordion-content-text"
    hidden=""
    id="radix-:r1:"
    role="region"
    style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
  />
</div>
`;

exports[`AccordionItem:  should render 2`] = `
<h3
  data-state="closed"
  data-testid="accordion-header"
>
  <button
    aria-controls="radix-:r1:"
    aria-expanded="false"
    class="group flex w-full flex-row items-center justify-between bg-transparent"
    data-radix-collection-item=""
    data-state="closed"
    data-testid="accordion-trigger"
    id="radix-:r0:"
    type="button"
  >
    <h3>
      item-title
    </h3>
    <svg
      aria-hidden="true"
      class="transition duration-300 group-data-[state=open]:rotate-180 group-data-[state=open]:transform"
      fill="currentColor"
      focusable="false"
      height="16"
      preserveAspectRatio="xMidYMid meet"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
      />
    </svg>
  </button>
</h3>
`;

exports[`AccordionItem:  should render 3`] = `
<div
  aria-labelledby="radix-:r0:"
  data-state="closed"
  data-testid="accordion-content-text"
  hidden=""
  id="radix-:r1:"
  role="region"
  style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
/>
`;

exports[`AccordionItem:  should render 4`] = `
<button
  aria-controls="radix-:r1:"
  aria-expanded="false"
  class="group flex w-full flex-row items-center justify-between bg-transparent"
  data-radix-collection-item=""
  data-state="closed"
  data-testid="accordion-trigger"
  id="radix-:r0:"
  type="button"
>
  <h3>
    item-title
  </h3>
  <svg
    aria-hidden="true"
    class="transition duration-300 group-data-[state=open]:rotate-180 group-data-[state=open]:transform"
    fill="currentColor"
    focusable="false"
    height="16"
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
    />
  </svg>
</button>
`;
